domain/machine/state/state.go

```diff
@@ -219,3 +219,77 @@ func (st *State) AllMachineNames(ctx context.Context) ([]machine.Name, error) {
 
 	return machineNames, nil
 }
+
+// RequireMachineReboot sets the reboot flag for the specified machine, which indicates that the
+// machine requires a reboot.
+//
+// Reboot flags are handled through the "machine_requires_reboot" table which contains only
+// machine UUID for which a reboot has been requested.
+func (st *State) RequireMachineReboot(ctx context.Context, uuid string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	setRebootFlag := `INSERT INTO machine_requires_reboot (machine_uuid) VALUES ($M.machine_uuid) ON CONFLICT DO NOTHING `
```

Comment:
`ON CONFLICT DO NOTHING` can bite us if we're not careful. We've decided to use it sparingly or not at all. I think the better approach is to handle the unique constraint error.

Reply:
Maybe we should discuss through it. It is a fairly simple table with only one column, so there is no risk to mess up with other column. AMHO it's the best way to achieve indempotency.

The issue to handle unique constraint error is that it implies either:
- import package `github.com/mattn/go-sqlite3` in this file, which may broke if `github.com/canonical/go-dqlite` changes its underlying binding
- update `github.com/canonical/go-dqlite` to alias errors from go-sqlite3 and  import package `github.com/canonical/go-dqlite` which may break if we change this dependency
- Alias in database package the required error (from sqlite3 or go-dqlite) and use here (maybe the least error-prone)
- Not use `errors.Is` to qualify the error (use pattern matching from error message), but it's error-prone.

I think `DO NOTHING` in this case, should bite **really hard** to worth this kind of complication.

Reply:
Ok, found the `database.IsErrConstraintPrimaryKey` which do the job in a way i am happy with :)